---
title: "spacemap basics"
author: "Christopher Conley, Pei Wang, Jie Peng"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{spacemap basics}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, echo = FALSE}
library(knitr)
opts_chunk$set(message=F, warning=F)
```

##Purpose 

This vignette will help users become familiar with the implementation of the model parameters and the output associated with the `spacemap` function, which is the fundamental model fitting function of the R package. Familiarity with the mathematical expression of the spaceMap model will help the reader map the notation to R object variables (see the Methods section of the spacemap publication).  

##Background 

spaceMap is a  conditional graphical model suited for learning networks from high-dimensional data constrained by limited sample size. The spaceMap model was motivated by integrative genomics applications to learn how genomic copy number alterations (CNA) predict and regulate the protein covariation response. It takes as input two distinct data types--predictors (.e.g. CNA)  and responses (e.g. proteins). The model can generally be applied to diverse high dimensional data types where interactions between predictors and response variables and also among responses are of primary importance, while interactions  among predictors is not of primary importance. Together, these interactions form a network and it is the goal of spaceMap to learn the network from predictor and response data. 

##Example 

Load the data set labeled `sim1`, which was generated under a hub network topology with two data types: predictor variables $\bf X \in \mathbb{R}^P$ and response variables $\bf Y \in \mathbb{R}^Q$, where $P = 14$ and $Q = 171$. When we fit the spaceMap model, we seek to learn the network comprised of node set $\textbf{V}= (X,Y)$ and edge set $\textbf{E}= \{ x_p - y_q : p \in \{1, \dots, P\}, q \in \{1, \dots, Q\} \} \cup \{ y_q - y_l : q \neq l \in \{1, \dots, Q\} \}$ 

```{r}
suppressPackageStartupMessages(library(spacemap))
data(sim1)
```

The `sim1` object contains the true network as well as data generated under the network topology.  Further details of `sim1` can be found by entering `help(spacemap::sim1)` at the R console. 

```{r, eval = FALSE, inlcude = FALSE, echo = FALSE, message=FALSE, warning=FALSE}
##In the interactive graphic below it can be seen there are 4 $X$ hubs depicted as large dark blue circles and a handful of $X$ variables with no edges at all depicted as small dark blue circles. There are two large disconnected components each containing 2 $X$ hubs and at least one $Y$ hub depicted as a large light blue circle. Other $Y$ variables are depicted as small light blue circles.
library(networkD3)
library(igraph)
info <- data.frame(id = c(sim1$Xindex, sim1$Yindex), 
                   type = c(rep(x = "x", times = length(sim1$Xindex)),
                            rep(x = "y", times = length(sim1$Yindex))),
                   hub = c(sim1$Xindex, sim1$Yindex) %in% sim1$XMasterIndex)
library(knitr)
info$id <- paste(toupper(info$type), info$id, sep = "-")
#set the diagonal to zero to avoid self-loops
tParCor <- sim1$trueParCor$yy
tGamma <- sim1$trueParCor$xy
diag(tParCor) <- 0
#the default convergence tolerance
tol <- 1e-6
tig <- spacemap::adj2igraph(YY = abs(tParCor) > tol, XY = abs(tGamma) > tol, info = info, dropnull = F)
#Convert the object suitable for `networkD3`
tig_d3 <- igraph_to_networkD3(tig, group = info$type)
#make the node size bigger according to degree
tig_d3$nodes$degree <- igraph::degree(tig)^2
forceNetwork(Links = tig_d3$links, Nodes = tig_d3$nodes, 
             #height = 700, width = 700,
             linkDistance = 5,
             Source = 'source', Target = 'target',
             fontSize = 12,
             NodeID = 'name', Group = 'group', Nodesize = "degree", 
             opacity = 0.90, 
             bounded = T, 
             charge = -15)
```

The spaceMap model learns the structure of the above network from data by estimating the edge set $\bf\hat E$. The following steps will discuss basic input parameters to learn the network. 


## Fitting `spacemap`

### Tuning penalties

Tuning penalties $\lambda_1, \lambda_2$ correspond to parameters `lam1` and `lam2`, respectively, where each determines the degree of overall sparsity of the $y-y$ and $x-y$ edges in the network. Tuning penalty $\lambda_3$ corresponds to parameter `lam3` and encourages selection of $x$ hubs--that is $x$'s that infuence many $y$'s. The selection of appropriate penalty parameters is of primary importance when evaluating `spacemap::spacemap` to learn a network. In this simplified scenario we set the tuning parameters based on previous cross validation results. The task of selecting appropriate tuning parameters requires a robust model-selection routine and is discussed in more advanced vignettes. 

```{r}
lam1 <- 69.67802 
lam2 <- 28.84801 
lam3 <- 12.38648
```

### Output

Fit the spacemap model to the data matrices $\bf X$ and $\bf Y$ with the selected tuning penalties. All other input parameters are left as defaults, but can be further customized as needed (see `help(spacemap::spacemap)`).  

```{r}
net <- spacemap(Y = sim1$Y, X = sim1$X, lam1 = lam1, lam2 = lam2, lam3 = lam3)
```

Most users wiil be interested in the network topology, which is encoded in the adjacency matrices `yy` and `xy` output from the `adjacency` auxillary function. 

```{r}
adjnet <- adjacency(net = net)
str(adjnet)
```

where 

+ `adjnet$yy[q,l]` indicates an edge $y_q - y_l$ when equal to 1 and 0 otherwise. 
+ `adjnet$xy[p,q]` indicates an edge $x_p - y_q$ when equal to 1 and 0 otherwise. 

The intermediate output of `spacemap` is a list of estimated parameters and model diagnostics. The estimated parameters are combined to infer the network. The model diagnostics help the user to decide whether the estimated parameters are valid.

#### Estimated parameters:

+ The estimated partial correlations $\bf \hat  \rho$ are reported as a symmetric matrix `ParCor` where the off-diagonal entries exceeding the convergence tolerance in absolute value encode $y-y$ edges. 
+ The estimated diagnoal of the inverse covariation matrix for the $y$ variables is denoted as $\bf \hat \theta$ and is reported as the numeric vector `sig.fit`. 
+ The estimated regression coefficients $\bf \hat \Gamma$ are reported in matrix form as `Gamma`, where  coefficients exceeding the convergence tolerance encode $x-y$ edges. 

#### Model Diagnostics: 

+ The residual sum of squares is reported as a numeric `rss`. 
+ The logical variable `convergence` indicates whether the model converged subject to the specified tolerance and number of maximum iterations; 
+ The numeric `deltaMax` reports the maximum $\Delta$ in parameter values in the last iteration. This helps diagnose how close 
the algorithm is to converging if it does not coverge. 

The structure of the intermediate output is shown below. 

```{r}
str(net)
```

## Quick Visualization

Often network analysis is enhanced by compelling visualizations that have nodes annotated with specific attributes like genomic ID,  genomic coordinates, and any pathways associated with the node. Annotating and visualizing network output from `spacemap` output is aided by several external and independent software libraries. The spacemap R package has auxillary functions to interact with the igraph R package, which contains a rich suite of network analysis tools. Below we illustrate how to quickly integrate annotation information into the network. 


```{r, eval = FALSE, echo = FALSE}
#NOT RUN
info <- data.frame(id = c(sim1$Xindex, sim1$Yindex), 
                   type = c(rep(x = "x", times = length(sim1$Xindex)),
                            rep(x = "y", times = length(sim1$Yindex))),
                   hub = c(sim1$Xindex, sim1$Yindex) %in% sim1$XMasterIndex)
library(knitr)
info$id <- paste(toupper(info$type), info$id, sep = "-")
kable(head(info))
```

Suppose you have nodes annotated in a data.frame where the rows are sorted to correspond to $(X_1, \dots, X_P, Y_1, \dots, Y_Q$). Suppose that predictor variables $X$ are genomic copy number aberations and response variables $Y$ are gene expression. Suppose that it has the following structure where 

```{r, echo = FALSE}
set.seed(15)
nx <- length(sim1$Xindex)
ny <- length(sim1$Yindex)
n <- nx + ny
xchr <- sample(x = 1:22, size = nx, replace = T)
ychr <- sample(x = 1:22, size = ny, replace = T)
chr <- c(xchr, ychr)
arm <- sample(x = c("p", "q"), size = nx, replace = T)
loc <- sample(x = 1:34, size = nx, replace = T)
xids <- paste0("chr", xchr, arm, loc)
yids <- paste0("gene", seq_len(ny))
begin <- sample(x = 1e7, size = n)
end <- sample(x = 2e7, size = n)
strand <- ifelse(begin < end, "+", "-")
pathways <- paste0("GO:", sprintf("%07d", sample(1e5, n)))
pathwayid <- replicate(ny, paste0(sample(x = pathways, size = sample(x = c(1,2,3),size = 1)), collapse = ";"))
type <- c(rep(x = "x", times = length(sim1$Xindex)),
                            rep(x = "y", times = length(sim1$Yindex)))
info <- data.frame(id = c(xids, yids),
                   node_level  = type,
                   chr = chr, 
                   begin = begin, 
                   end = end, 
                   strand = strand, 
                   pathways = c(rep(NA, nx), pathwayid), 
                   stringsAsFactors = F, row.names = c(xids, yids))
kable(info[10:20,], row.names = FALSE)
```

Load the node annotation data.frame.  

```{r, eval = F}
if(0) { 
  info <- readRDS(file = "/path/to/my/annotation/file.rds")
}
```

Convert the spacemap output to igraph format together with the annotation information. 

```{r}
library(igraph)
ig <- spacemap::adj2igraph(YY = adjnet$yy, XY = adjnet$xy, info = info, dropnull = F)
```

The `igraph` object has a very convenient sparse representation as shown below. 

```{r}
ig
```


Identify modules of the inferred network with `igraph`'s edge-betweenness algorithm. Each module discovered will be colored differently when we visualize the network. 

```{r}
ceb <- igraph::cluster_edge_betweenness(graph = ig)
members <- igraph::membership(ceb)
```


While there are many visualization options, we recommend using Cytoscape (not part of R) to visually inspect networks. However, Cytoscape may involve some manual curation and is a sophisticated tool with a learning curve. If you simply need a quick way to visualize the network output, the following section illusrates how to do so within R.  For interactive visualization, we may make use of the `networkD3` package. The following snippet will visualize `spacemap`'s network where $X$ and $Y$ hub nodes are larger in size. Colors denote the 14 different modules of the network. Hover over a node with your mouse if you wish to see the node ID and its adjacent neighbors. 

 
```{r}
library(networkD3)
#Convert the object suitable for `networkD3`
ig_d3 <- igraph_to_networkD3(ig, group = members)
#make the node size bigger according to degree
ig_d3$nodes$degree <- igraph::degree(ig)^2
forceNetwork(Links = ig_d3$links, Nodes = ig_d3$nodes, 
             linkDistance = 10,
             Source = 'source', Target = 'target',
             fontSize = 12,
             NodeID = 'name', Group = 'group', Nodesize = "degree", 
             opacity = 0.90, 
             bounded = T, 
             charge = -15)
```


## Further Reading

With this basic understanding of the `spacemap` model fitting, we recommend looking at the [next vignette](https://topherconley.github.io/spacemap/articles/tuning.html) which illustrates how to select tuning penalty parameters through an iterative grid-search with cross-validation criterion. 

